### ISSUE TITLE
Immediate open() on VirtualComboBox causes errors

### ISSUE THREAD

**Issue by mbgonicus** [qooxdoo/qooxdoo#9361]
[Example in Playground](http://tinyurl.com/ydctvo5d)

In the example there is a `qx.ui.form.VirtualComboBox` that I like to open initially. Trying to call `open()` on it results in: `GlobalError: Cannot read property 'width' of null`

Depending on whether it was added to a container or not, the drop down will open on different places. Try the commented calls in the playground.

A workaround is to defer the call of `open`.

**Comment by level420** [comment]
@mbgonicus doesn't this simply mean that you have to wait until the the combo box is rendered? How about simply using a "appear" event listener to open the combo box initially? http://tinyurl.com/ybdywjvk

**Comment by level420** [comment]
as an alternative you could also flush the queue before calling open: http://tinyurl.com/y9o2zbbk

**Comment by mbgonicus** [comment]
That is correct, but I consider this of being a workaround. If one cannot open a VirtualComboBox before it is being rendered, I reckon it should be at least clearly documented. But imho it's in the responsibility of the VirtualComboBox, i.e. if one calls open() and it is not rendered, it defers the call transparently by itself. At least this behaviour was very surprising to me.

**Comment by level420** [comment]
The idea behind all this is to have a rendering pipeline which first is filled by widget instantiation and dom creation/manipulation workload and is then flushed, normally when the current javascript code block finishes. There are a lot of widgets which are not able to successfully call some methods until they are not fully rendered, thus the queue did not complete yet.

A lot of code in qooxdoo widgets, for example the child widget creation, is postponed (lazy instantiation) until that widget is needed. In case of `qx.ui.form.VirtualComboBox` the open method fails because the code which calculates and adapts the dimensions of the drop down list `qx.ui.form.core.VirtualDropDownList` fails, because the virtual combo box itself is not fully rendered yet. This is in the private method `__adjustWidth` at https://github.com/qooxdoo/qooxdoo/blob/master/framework/source/class/qx/ui/form/core/VirtualDropDownList.js#L423 There `this._target.getBounds()` is called before it is rendered, failing with being unable to get `width` of null.

One possibility would be to change all methods which potentially could be called before the DOM of a widget is created (and added to its parent, an got its right dimension and position) to flush the queue if the DOM was not created yet. This can get complicated for a widget like `qx.ui.form.VirtualComboBox` which itself is composed by multiple sub widgets.

The second option is to wait until all is de-queued and listen on the appear event.

In this special case, we could at least check if we currently are able to adapt the drop down list to the combo box by testing if `this._target.getBounds()` gets a non null result and postpone the adjustment until the `this._target` sends its `appear` event with a `addListenerOnce`.

Or we could add some code postponing the `open` itself if the DOM does not exist yet, or flush the queues to be able to open the popup.

**Comment by level420** [comment]
@mbgonicus As allways PRs are very welcome. This could not only be code PRs but documenation PRs as well.

**Comment by mbgonicus** [comment]
When time allows it, I'll try a pr

**Comment by derrell** [comment]
I agree that all widgets should just work without having to resort to flushing queues or adding appear listeners. Those are kludges to make work what should have worked all along. Those of us who have been qxusers for a long time have gotten so used to doing them that we don't really even think about it any longer, but it certainly does add to the already very steep learning curve for qooxdoo. Fixing this would be wonderful, but is likely to be a lot of work.

**Comment by johnspackman** [comment]
@level420 is it definitely safe to flush the queues at any time?  ie recursion would not be a problem?

IMHO the second option would be preferable - is this really such a lot of work?  because it looks fairly limited to `__adjustWidth` at first glance

**Comment by derrell** [comment]
@johnspackman It's not a lot of work to fix it in one place. It's a lot of work to fix it globally throughout qooxdoo, which I believe should be done when someone frees up the time to accomplish it.

**Comment by derrell** [comment]
Then again, we could just fix it piecemeal as someone asks about a particular case of "why in the world do I have to do this?"

**Comment by level420** [comment]
@johnspackman the definitely safest way would be to wait until the `appear`event is fired.

Regarding the code fix in `__adjustWidth`: if we fix that it is still not what is really needed, as the `open` then succeeds, but opens the pulldown list of the combobox somewhere "in the green" because the pulldown list adapts to the input field, which maybe is not fully rendered and positioned yet.

The only thing what would help here is to check if the widget is not rendered yet and postpone the open call for after rendering is done. Pseudocode for the open method of the combobox

```javascript
   open : function() {
     if(!this.rendered) {
      this.addListenerOnce("appear", this.open, this);
     }
     
     // do watever is needed to really open
   }
```

**Comment by johnspackman** [comment]
i didn't realise it was so widespread - I think I've only had to explicitly flush once or twice, ever.  I've just checked my code and can only find two references to `qx.ui.core.queue.Manager.flush`, one is a very old component I dont use any more and the other is commented out...

Does this particularly effect specific parts of the framework only, eg the virtual stuff?

But yes, I think encouraging people to fix piecemeal is a good idea (better than nothing)

**Comment by oetiker** [comment]
@level420 and @johnspackman I also use the appear pattern quite often ... works well ... but obviously it requires one to understand the way qx works ... as soon as other libraries are involved which operate on DOM elements, this becomes a very important concept ... it is certainly fixable 'inside' the qx universe, but as soon as other js stuff is involved, one has to understand how things work ... 

In general, I don't think excessive flushing is ideal ... as this tends to  slow things down.

**Comment by derrell** [comment]
I find myself doing things in "appear" listeners frequently. For example, adding an item to the end of a list and then scrolling to the end (think chat app where a new message was just received from your chat partner) requires waiting on "appear" before scrolling to the end. That shouldn't be necessary. You should be able add the list item, utter scrollToEnd(), and qooxdoo should handle waiting to scroll until the queue has been flushed. This sort of thing happens frequently.

**Comment by derrell** [comment]
And I wouldn't propose fixing this with flushes; rather with internally handling the "appear" event when necessary.

**Comment by level420** [comment]
But from nature such internally delayed methods are asynchronous, where the user also needs to understand how qooxdoo work. You could, taking the example in this issue, then NOT rely on the combo box being open after you call `cb.open()`. Then you'll need an event which fires after the combo box is really open.

**Comment by level420** [comment]
We could of course raise an exception like `widget not rendered yet` if someone calls `open`.

**Comment by cboulanger** [comment]
Or, as part of using new language features, open() could return a Promise (I haven't checked but assume that it doesn't return anything yet).

```javascript
myVirtualComboBox.open().then( () =>{
  // do something with myVirtualComboBox
})
```

**Comment by cboulanger** [comment]
This could be visibly documented and would fit to modern async expectations - although not with the traditional "qooxdoo ways". But I am very much for gradually integrating the new JS-Features.

**Comment by mbgonicus** [comment]
It seems that there is no clear consensus how to deal with this in general, thus I fixed it for this very case and made the PR #9371.

**Comment by level420** [comment]
Fixed via https://github.com/qooxdoo/qooxdoo/pull/9371
